### PR TITLE
Default values in procedures

### PR DESCRIPTION
**Pull request by pontusmelke**
Provide default values in procedures:

Syntax: 

```
@Procedure
public Stream<Output> simpleArgumentWithDefault( @Name( value = "name", defaultValue = "42") long someValue )
{
    return Stream.of( new Output( someValue ) );
}
```

The default value is given as the corresponding cypher representation of the given type.
For example:
- `@Name( value = "name", defaultValue = "null") Map<String,Object> someValue`
- `@Name( value = "name", defaultValue = "{foo: ['bar', 'baz']}") Map<String,Object> someValue`
- `@Name( value = "name", defaultValue = "[42, 1337]") List<Long> someValue`

changelog: [procedures] Support for default parameters in procedures,

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ParseUtil.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+                break;
+            case '\'':
+                closing = s.indexOf( '\'', pos + 1 );
+                if ( closing < 0 )
+                {
+                    throw new IllegalArgumentException( "Did not find a matching end quote, '" );
+                }
+                return s.substring( pos + 1, closing );
+            case '\"':
+                closing = s.indexOf( '\"', pos + 1 );
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ParseUtil.java]: how do we handle escaped quotes (e.g., "a\"bc")?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ParseUtil.java [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        {
+            char c = s.charAt( pos );
+            int closing;
+            switch ( c )
+            {
+            case ' ':
+                ++pos;
+                break;
+            case '\'':
+                closing = s.indexOf( '\'', pos + 1 );
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ParseUtil.java]: how do we handle escaped quotes (e.g., 'a\'bc')?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ListConverter.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        this.type = type;
+        this.neoType = neoType;
+    }
+
+    @Override
+    public Neo4jValue apply( String s )
+    {
+        String value = s.trim();
+        if ( value.equalsIgnoreCase( "null" ) )
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ListConverter.java:46]: why not using an actual `null` value instead the string "null"?

File: community/kernel/src/test/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureWithArgumentsTest.java [hunk 4]
```diff
@@ -156,6 +182,34 @@ public MyOutputRecord( String name )
         }
     }
 
+    public static class ClassWithProcedureWithDefaults
+    {
+        @Procedure
+        public Stream<MyOutputRecord> defaultValues( @Name( value = "a", defaultValue = "a") String a , @Name( value = "b", defaultValue = "42") long b, @Name( value = "c",
+                defaultValue = "3.14") double c)
```
Comment (davidegrohmann) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureWithArgumentsTest.java:189]: why the default value should be always a string that we need to parse?

Comment (pontusmelke) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureWithArgumentsTest.java:189]: It is the unfortunate behaviour of annotations in java, you can only provide `compile time constants`, so you cannot do `defaultValue = Collections.emptyList()`, or even `defaultValue = null`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ListConverter.java (hunk 3 above)
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ListConverter.java:46]: It needs to be a _compile time constant_, and for some reason Java does not think it is, http://stackoverflow.com/questions/17758796/why-isnt-null-a-compile-time-constant

**Comment by pontusmelke** [comment]
@davidegrohmann Good catch on the escaped quotes. I would love to replace the rather naive parser with something else but it doesn't feel like we could add a dependency on something like Jackson in `kernel`? However since it doesn't need to be ultra fast as the strings are only parsed once at boot time - and I don't expect huge json-documents in the annotations - I think this is enough.

**Comment by boggle** [comment]
I like this feature, it is really nice. I'm a bit concerned about using a native parser though, it makes it harder to ensure that both Cypher and the native parser agree on what a valid literal Cypher value is. I'd prefer us to use our own parboiled parser for that effect and since execution engine is being injected into kernel anyhow that shouldn't be too hard.

**Comment by davidegrohmann** [comment]
@pontusmelke yeah I share your concerns, I am actually wondering how can we differentiate between "42" the number and "42" the string

**Comment by pontusmelke** [comment]
@boggle It is not a full Cypher parser, it is a simple map and list parser. I wouldn't want to use the parboiled parser from Java, oh you mean get hold of the ExecutionEngine and basically do `engine.run`?
@davidegrohmann it is given from the context, say you have

```
@Procedure
void proc(@Name(value = "value", defaultValue = "42") long value)
...
```

Here we will use the _integer parser_ which is basically just `Long.parseLong`, if it would have been `String value` we would just pass the value through.  The  `MapConverter` is only used for procedures that  declare that they take a `Map<String, Object>`, and the `ListConverter` only for say `List<Long>`.

https://github.com/pontusmelke/neo4j/blob/a72b8ec20adf977fb0b5d1aa65f6fd1a613a3072/community/kernel/src/main/java/org/neo4j/kernel/impl/proc/TypeMappers.java#L141
https://github.com/pontusmelke/neo4j/blob/a72b8ec20adf977fb0b5d1aa65f6fd1a613a3072/community/kernel/src/main/java/org/neo4j/kernel/impl/proc/TypeMappers.java#L154

@boggle @davidegrohmann I would like to change `value` to `name` so that it is

```
@Procedure
void proc(@Name(name = "value", defaultValue = "42") long value)
...
```

instead. Now when you use named parameters I feel that `value` and `defaultValue` is confusing. WDYT?

Nevermind, `value` is a magic annotation thing

**Pull request merged**